src/main/java/io/github/seggan/sfcalc/Calculator.java

```diff
@@ -100,66 +105,71 @@ public void printResults(@Nonnull CommandSender sender, @Nonnull SlimefunItem it
             if (item == null || item.getType().isAir()) {
                 continue;
             }
-
             add(inv, item, item.getAmount());
         }
 
         return inv;
     }
 
     @Nonnull
-    public Map<ItemStack, Long> calculate(@Nonnull SlimefunItem item) {
-        Map<ItemStack, Long> result = new HashMap<>();
-        Map<ItemStack, Map<ItemStack, Long>> calculated = new HashMap<>(); // stores names that are already calculated for reference
+    public Map<ItemStack, Long> calculate(@Nonnull SlimefunItem parent, Long amount) {
 
-        for (ItemStack i : item.getRecipe()) {
-            if (i == null) {
-                // empty slot
-                continue;
-            }
+        Map<ItemStack, Long> result = new HashMap<>();
+        add(result, parent.getItem(), amount);
 
-            int amount = i.getAmount();
+        //uncraft the material
+        add(result, parent.getItem(), -parent.getRecipeOutput().getAmount());
+        for(ItemStack item : parent.getRecipe()) {
+            if(item == null) continue;
+            add(result, item, item.getAmount());
+        }
 
-            if (calculated.containsKey(i)) { // check already calculated items
-                addAll(result, calculated.get(i), amount);
-                continue;
+        //uncraft submaterials
+        SlimefunItemStack next = getNextItem(result);
+        while(next != null) {
+            int multiplier = next.getItem().getRecipeOutput().getAmount();
+            Long operations = (result.get(next)+multiplier-1)/multiplier; //ceiling(needed/multiplier) but abusing fast ints
```

Comment:
```suggestion
            long operations = (result.get(next)+multiplier-1)/multiplier; //ceiling(needed/multiplier) but abusing fast ints
```
please, dont use wrapped types